src/libraries/System.Private.CoreLib/src/System/IO/FileStreamOptions.cs

```diff
@@ -5,32 +5,139 @@ namespace System.IO
 {
     public sealed class FileStreamOptions
     {
+        private FileMode? _mode;
+        private FileAccess? _access;
+        private FileShare _share = FileStream.DefaultShare;
+        private FileOptions _options;
+        private long _preallocationSize;
+        private int _bufferSize = FileStream.DefaultBufferSize;
+
         /// <summary>
         /// One of the enumeration values that determines how to open or create the file.
         /// </summary>
-        public FileMode Mode { get; set; }
+        /// <exception cref="T:System.ArgumentOutOfRangeException">When <paramref name="value" /> contains an invalid value.</exception>
+        /// <exception cref="T:System.ArgumentException">When <paramref name="value" /> is <see cref="System.IO.FileMode.Append" /> but <see cref="Access" />  is not <see cref="System.IO.FileAccess.Write" />.
+        ///  -or-
+        /// <see cref="Access" /> is <see cref="System.IO.FileAccess.Read" /> while given mode requires <see cref="System.IO.FileAccess.Write" />.
+        /// </exception>
+        public FileMode Mode
+        {
+            get => _mode.HasValue ? _mode.GetValueOrDefault() : FileMode.Open;
+            set
+            {
+                if (value < FileMode.CreateNew || value > FileMode.Append)
+                {
+                    ThrowHelper.ArgumentOutOfRangeException_Enum_Value();
+                }
+                else if (_access.HasValue && (_access & FileAccess.Read) != 0 && value == FileMode.Append)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidAppendMode, nameof(value));
+                }
+                else if (_access.HasValue && (_access & FileAccess.Write) == 0)
+                {
+                    if (value == FileMode.Truncate || value == FileMode.CreateNew || value == FileMode.Create || value == FileMode.Append)
+                    {
+                        throw new ArgumentException(SR.Format(SR.Argument_InvalidFileModeAndAccessCombo, value, _access), nameof(value));
+                    }
+                }
```

Comment:
Hmm.  We typically consider validation that crosses properties like this to be problematic.  I hadn't noticed the interactions when I suggested moving the validation logic into the setters.  @bartonjs, opinions?

Reply:
(A happy medium might be leaving in FileStreamOptions the validation that depends only on the value being set, and put back in FileStream the rest, e.g. here we'd keep the `if (value < FileMode.CreateNew || value > FileMode.Append)` check but remove the else-ifs.  The validation for Share/Options/PreallocationSize/BufferSize is all fine as it's based only on the value being set.)

Reply:
@stephentoub done

Reply:
The answer may no longer be needed, but since the question was asked:

The guidelines are that property sets shouldn't interfere with each other, but that any validation wait until a method depends on the state of those properties.

So the first if "FileMode makes no sense" makes sense here, but the remainder shouldn't happen until something wants to consume one.

If you want to keep the validation succinct, and in this type, you could add an internal `Validate` method, that way the same validation is shared across all consumers.